File: shared/src/main/java/ru/privatenull/pnauth/dialog/DialogForm.java
package ru.privatenull.pnauth.dialog;

import net.kyori.adventure.text.Component;
import ru.privatenull.pnauth.platform.PnPlayer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.function.Consumer;

/**
 * High-level native form API. Transport identifiers are generated per display and button
 * responses are routed directly to their handlers.
 */
public final class DialogForm {
    private final DialogLayout layout;
    private final List<Button> buttons;
    private final int columns;
    private final Consumer<DialogResponse> closeHandler;

    private DialogForm(Builder builder) {
        this.layout = new DialogLayout(builder.title, builder.externalTitle, builder.body, builder.inputs,
                builder.canCloseWithEscape, builder.pause, builder.afterAction);
        this.buttons = List.copyOf(builder.buttons);
        this.columns = builder.columns;
        this.closeHandler = builder.closeHandler;
    }

    public static Builder builder(Component title) {
        return new Builder(title);
    }

    DialogHandle show(PlayerDialogs dialogs, PnPlayer player) {
        Objects.requireNonNull(dialogs, "dialogs");
        Objects.requireNonNull(player, "player");
        String token = UUID.randomUUID().toString().replace("-", "");
        Map<String, Consumer<DialogResponse>> handlers = new LinkedHashMap<>();
        List<DialogButton> materialized = new ArrayList<>(buttons.size());
        for (int index = 0; index < buttons.size(); index++) {
            Button button = buttons.get(index);
            String actionId = "pnauth:form_" + token + "_" + index;
            handlers.put(actionId, button.handler);
            materialized.add(new DialogButton(button.label, button.tooltip, button.width,
                    new DialogAction.DynamicCustom(actionId, Map.of())));
        }
        PlayerDialog dialog = new PlayerDialog("pnauth:form_" + token, layout,
                new DialogType.MultiAction(materialized, null, columns));
        DialogHandle handle = dialogs.show(player, dialog);
        handle.onResponse(response -> {
            if (response.closed()) {
                if (closeHandler != null) closeHandler.accept(response);
                return;
            }
            Consumer<DialogResponse> handler = handlers.get(response.action());
            if (handler != null) handler.accept(response);
        });
        return handle;
    }

    private record Button(Component label, Component tooltip, int width, Consumer<DialogResponse> handler) { }

    public static final class Builder {
        private final Component title;
        private Component externalTitle;
        private final List<DialogBody> body = new ArrayList<>();
        private final List<DialogInput> inputs = new ArrayList<>();
        private final List<Button> buttons = new ArrayList<>();
        private boolean canCloseWithEscape = true;
        private boolean pause;
        private DialogLayout.AfterAction afterAction = DialogLayout.AfterAction.CLOSE;
        private int columns = 1;
        private Consumer<DialogResponse> closeHandler;

        private Builder(Component title) {
            this.title = Objects.requireNonNull(title, "title");
        }

        public Builder externalTitle(Component value) { externalTitle = value; return this; }

        public Builder body(Component content, int width) {
            body.add(new DialogBody.PlainMessage(Objects.requireNonNull(content, "content"), width));
            return this;
        }

        public Builder input(DialogInput input) {
            inputs.add(Objects.requireNonNull(input, "input"));
            return this;
        }

        public Builder text(String id, Component label, int maximumLength, int width) {
            return input(new DialogInput.Text(id, label, true, "", maximumLength, width, null));
        }

        public Builder toggle(String id, Component label, boolean initialValue) {
            return input(new DialogInput.Toggle(id, label, initialValue, null, null));
        }

        public Builder button(Component label, Consumer<DialogResponse> handler) {
            return button(label, Component.empty(), 150, handler);
        }

        public Builder button(Component label, Component tooltip, int width,
                              Consumer<DialogResponse> handler) {
            buttons.add(new Button(Objects.requireNonNull(label, "label"),
                    tooltip == null ? Component.empty() : tooltip, width,
                    Objects.requireNonNull(handler, "handler")));
            return this;
        }

        public Builder onClose(Consumer<DialogResponse> handler) { closeHandler = handler; return this; }
        public Builder columns(int value) { columns = value; return this; }
        public Builder canCloseWithEscape(boolean value) { canCloseWithEscape = value; return this; }
        public Builder pause(boolean value) { pause = value; return this; }
        public Builder afterAction(DialogLayout.AfterAction value) { afterAction = value; return this; }

        public DialogForm build() {
            if (buttons.isEmpty()) throw new IllegalStateException("A form must contain at least one button");
            if (columns < 1) throw new IllegalStateException("Form columns must be positive");
            return new DialogForm(this);
        }
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/dialog/DialogForm.kt
package ru.privatenull.pnauth.dialog

import net.kyori.adventure.text.Component
import ru.privatenull.pnauth.platform.PnPlayer
import java.util.UUID
import java.util.function.Consumer

/**
 * High-level native form API. Transport identifiers are generated per display and button
 * responses are routed directly to their handlers.
 */
class DialogForm private constructor(builder: Builder) {
    val layout: DialogLayout = DialogLayout(
        builder.title, builder.externalTitle, builder.body, builder.inputs,
        builder.canCloseWithEscape, builder.pause, builder.afterAction
    )
    val buttons: List<Button> = builder.buttons.toList()
    val columns: Int = builder.columns
    val closeHandler: Consumer<DialogResponse>? = builder.closeHandler

    internal fun show(dialogs: PlayerDialogs, player: PnPlayer): DialogHandle {
        val token = UUID.randomUUID().toString().replace("-", "")
        val handlers = LinkedHashMap<String, Consumer<DialogResponse>>()
        val materialized = ArrayList<DialogButton>(buttons.size)
        for (index in buttons.indices) {
            val button = buttons[index]
            val actionId = "pnauth:form_${token}_$index"
            handlers[actionId] = button.handler
            materialized.add(
                DialogButton(
                    button.label, button.tooltip, button.width,
                    DialogAction.DynamicCustom(actionId, emptyMap())
                )
            )
        }
        val dialog = PlayerDialog(
            "pnauth:form_$token", layout,
            DialogType.MultiAction(materialized, null, columns)
        )
        val handle = dialogs.show(player, dialog)
        handle.onResponse { response ->
            if (response.closed) {
                closeHandler?.accept(response)
                return@onResponse
            }
            val handler = handlers[response.action]
            handler?.accept(response)
        }
        return handle
    }

    data class Button(
        val label: Component,
        val tooltip: Component,
        val width: Int,
        val handler: Consumer<DialogResponse>
    )

    class Builder(val title: Component) {
        var externalTitle: Component? = null
            private set
        val body: MutableList<DialogBody> = ArrayList()
        val inputs: MutableList<DialogInput> = ArrayList()
        private val _buttons: MutableList<Button> = ArrayList()
        internal val buttons: List<Button> get() = _buttons
        var canCloseWithEscape: Boolean = true
            private set
        var pause: Boolean = false
            private set
        var afterAction: DialogLayout.AfterAction = DialogLayout.AfterAction.CLOSE
            private set
        var columns: Int = 1
            private set
        var closeHandler: Consumer<DialogResponse>? = null
            private set

        fun externalTitle(value: Component?): Builder {
            externalTitle = value
            return this
        }

        fun body(content: Component, width: Int): Builder {
            body.add(DialogBody.PlainMessage(content, width))
            return this
        }

        fun input(input: DialogInput): Builder {
            inputs.add(input)
            return this
        }

        fun text(id: String, label: Component, maximumLength: Int, width: Int): Builder {
            return input(DialogInput.Text(id, label, true, "", maximumLength, width, null))
        }

        fun toggle(id: String, label: Component, initialValue: Boolean): Builder {
            return input(DialogInput.Toggle(id, label, initialValue, null, null))
        }

        fun button(label: Component, handler: Consumer<DialogResponse>): Builder {
            return button(label, Component.empty(), 150, handler)
        }

        fun button(
            label: Component,
            tooltip: Component?,
            width: Int,
            handler: Consumer<DialogResponse>
        ): Builder {
            _buttons.add(Button(label, tooltip ?: Component.empty(), width, handler))
            return this
        }

        fun onClose(handler: Consumer<DialogResponse>?): Builder {
            closeHandler = handler
            return this
        }

        fun columns(value: Int): Builder {
            columns = value
            return this
        }

        fun canCloseWithEscape(value: Boolean): Builder {
            canCloseWithEscape = value
            return this
        }

        fun pause(value: Boolean): Builder {
            pause = value
            return this
        }

        fun afterAction(value: DialogLayout.AfterAction): Builder {
            afterAction = value
            return this
        }

        fun build(): DialogForm {
            check(_buttons.isNotEmpty()) { "A form must contain at least one button" }
            check(columns >= 1) { "Form columns must be positive" }
            return DialogForm(this)
        }
    }

    companion object {
        @JvmStatic
        fun builder(title: Component): Builder {
            return Builder(title)
        }
    }
}
